lldb/include/lldb/Symbol/CompilerType.h

```diff
@@ -194,6 +192,54 @@ class CompilerType {
   bool IsTypedefType() const;
 
   bool IsVoidType() const;
+
+  bool IsSmartPtrType() const;
+
+  bool IsInteger() const;
+
+  bool IsFloat() const;
+
+  bool IsEnumerationType() const;
+
+  bool IsUnscopedEnumerationType() const;
+
+  bool IsIntegerOrUnscopedEnumerationType() const;
+
+  bool IsSigned() const;
+
+  bool IsNullPtrType() const;
+
+  bool IsBoolean() const;
+
+  bool IsEnumerationIntegerTypeSigned() const;
+
+  bool IsScalarOrUnscopedEnumerationType() const;
+
+  bool IsPromotableIntegerType() const;
+
+  bool IsPointerToVoid() const;
+
+  bool IsRecordType() const;
+
+  bool IsVirtualBase(CompilerType target_base, CompilerType *virtual_base,
+                     bool carry_virtual = false) const;
+
+  bool IsContextuallyConvertibleToBool() const;
+
+  bool IsBasicType() const;
+
+  std::string TypeDescription();
+
+  bool CompareTypes(CompilerType rhs) const;
+
+  const char *GetTypeTag();
+
+  uint32_t GetNumberOfNonEmptyBaseClasses();
+
+  CompilerType GetTemplateArgumentType(uint32_t idx);
+
+  CompilerType GetSmartPtrPointeeType();
```

Comment:
There seems to be some overlap between this and the C++ shared_ptr synthetic child provider. I think it would be good if either the formatter uses this functionality or if the intended user of this function uses the synthetic children instead.

Reply:
Could you elaborate on this a little bit more please. I'm still learning the LLDB type system, and I don't quite understand what you're asking me to do.

Reply:
Here you are adding a function that takes a CompilerType holding an STL  smart pointer and ask for the pointee type. We also have synthetic child formatters used by `frame variable` that allow evaluating `*my_shared_ptr`. From that I was assuming that there must be some overlap between what the dataformatter provides and what this function does.
What I don't know is whether synthetic children provided by a formatter are a good fit for what you are trying to do, but maybe the inverse is true, and the existing data formatter should be implemented in terms of this new GetSmartPtrPointeeType() API.
I just want to make sure we're not introducing two places in LLDB that have to know about STL smart pointers while covering slightly different functionality.

Reply:
Note that this function returns the type of the shared pointer, whereas the synthetic child providers return values of that type.  So if anything you would implement the formatter on top of this type recognition (though that only gets you a little way there, you also have to find where the value is in the smart pointer.)

It's a little awkward that these API's speak of "smart pointer" when they really mean "the system library provided memory managing smart pointer".  But I can't think of a good name to make this explicit.

However, that brings us round to the synthetic child support for "smart pointers".  That is a feature that's not limited to the standard library's notion of a smart pointer.  It really just tells lldb how to interpret the `->` in:

(lldb) frame var foo->bar

The implementation of this feature doesn't require any knowledge of what the arrow operator for foo does (it is not even required to have one).  For instance, you could use it to have lldb log uses of the `->` operator for this type in the expression parser, or whatever.  The only requirement is that you define a synthetic child provider that emulates the -> operator, returning a ValueObject from the appropriate API in the provider.

So at some point someone could redo the STL smart pointer comprehension in lldb to use the synthetic child arrow operator emulation, that's a much broader feature that what this API is promising.

Reply:
Do you want me to try to update the data formatter to you my function in this PR, or should we commit this PR without that change, with the promise that I will submit a follow-up PR with the data formatter change?

Jim, I follow what you're saying but I'm not sure if there's a concrete action you're asking me to take?

Reply:
> Do you want me to try to update the data formatter to you my function in this PR, or should we commit this PR without that change, with the promise that I will submit a follow-up PR with the data formatter change?

If it's easy to do, sure — but I'm also fine with postponing it for a follow-up commit. I just want to make sure we have a clear plan for how to make that happen.

Reply:
I would prefer to do it in a follow-up commit. (My next PR!).

Reply:
I was mostly trying to say that the ValueObject support for "faking the arrow operator effects" is orthogonal to having the type system contain the notion of "smart pointers".  

I also think the whole naming of this is confusing.  What is a "smart pointer"?  The IsSmartPtr states some intent about this:

`  /// This determines if the type is a shared, unique or weak pointer, either
  /// from stdlibc++ or libc+++.
  bool IsSmartPtrType() const;
`

But that is also a little odd.  This is a general type system query but doesn't in fact have anything to do with the type system.  Rather with a particular implementation in the C++ STL - which is orthogonal to the C++ type system...  So it seems like an awfully C++ specific API to have in the Typesystem.

Reply:
Having a comment saying "This is defined only for C++ types" in a function from CompilerType.h seems wrong.

Reply:
I can remove these helper functions from the CompilerType class and put them into my parser class for DIL instead, if you think that would be better.

Reply:
Clarification: I meant "I can remove the helper functions related to smart pointers."